src/Microsoft.VisualStudio.Composition/Configuration/MetadataViewGenerator.cs

```diff
@@ -94,12 +96,34 @@ internal static class MetadataViewGenerator
 
         public delegate object MetadataViewFactory(IReadOnlyDictionary<string, object?> metadata, IReadOnlyDictionary<string, object?> defaultMetadata);
 
-        private static AssemblyBuilder CreateProxyAssemblyBuilder()
+        private static AssemblyBuilder CreateProxyAssemblyBuilder(ImmutableHashSet<AssemblyName> assemblies)
         {
-            var proxyAssemblyName = new AssemblyName(string.Format(CultureInfo.InvariantCulture, "MetadataViewProxies_{0}", Guid.NewGuid()));
+            var proxyAssemblyName = new AssemblyName(string.Format(CultureInfo.InvariantCulture, "MetadataViewProxies_{0}", GenerateGuidFromAssemblies(assemblies)));
             return AssemblyBuilder.DefineDynamicAssembly(proxyAssemblyName, AssemblyBuilderAccess.Run);
         }
 
+        private static Guid GenerateGuidFromAssemblies(ImmutableHashSet<AssemblyName> assemblies)
+        {
+            // To make dll load and JIT comparisons between builds easier, we generate a
+            // consistent name for the proxy assembly based on the input assemblies involved.
+            using var algorithm = SHA256.Create();
+
+            foreach (AssemblyName name in assemblies.OrderBy(a => a.Name, StringComparer.Ordinal))
+            {
+                var bytes = Encoding.UTF8.GetBytes(name.Name!);
```

Comment:
Why take the time to re-encode UTF16 bytes as UTF8? What if we just take the pointer to the UTF16 string we already have and encode the bytes in that memory directly? That would be faster.